test/semantics/select-case.f90

```diff
@@ -0,0 +1,172 @@
+   real :: grade4 = 2.0
+   logical, parameter :: grade5 = .false.
+   CHARACTER(KIND=1), PARAMETER    :: ze = 'a', at='b'
+   CHARACTER(KIND=4), PARAMETER :: mh='c'
+
+   select case (grade1)
+      case ('A')
+      print*, "Excellent!"
+      case ('B')
+      case ('C')
+         print*, "Well done"
+      case default
+         print*, "Invalid grade"
+   end select
+
+   select case (grade2)
+      case (1)
+      print*, "Excellent!"
+      case (2)
+      case (3)
+         print*, "Well done"
+      case default
+         print*, "Invalid grade"
+   end select
+
+   select case (grade3)
+      case (.true.)
+       name = 'true'
+      case (.false.)
+        name = 'false'
+   end select
+   print*, "Your grade is ", name
+
+   select case (name)
+      case default
+         print*, "Invalid grade"
+      case ('now')
+      case ('test')
+         print*, "Well done"
+   end select
+
+   !ERROR: SELECT CASE expression must be of type CHARACTER, INTEGER, OR LOGICAL
+   select case (grade4)
+      case (1.0)
+      print*, "Excellent!"
+      case (2.0)
+      case (3.0)
+         print*, "Well done"
+      case default
+         print*, "Invalid grade"
+   end select
+
+   select case (grade3)
+      case default
+         print*, "Invalid grade"
+      case (.true.)
+         print*, "Well done"
+      !ERROR: Not more than one of the selectors of SELECT CASE statement may be default
+      case default
+         print*, "Invalid grade"
+   end select
+
+   select case (grade3)
+      case default
+         print*, "Invalid grade"
+      case (.true.)
+      !ERROR: SELECT CASE value must be of type LOGICAL
+      case (3)
+         print*, "Well done"
+   end select
+
+   select case (grade3)
+      case default
+         print*, "Invalid grade"
+      !ERROR: SELECT CASE expression of type LOGICAL must not have range of case value
+      case (.true. :)
+   end select
+
+   select case (grade2)
+      case default
+         print*, "Invalid grade"
+      case (2 :)
+         print*, "Well done"
+      !ERROR: SELECT CASE value must be of type INTEGER
+      case (.true. :)
+   end select
+
+   select case (grade3)
+      case (.true.)
+       name = 'true'
+      case (.false.)
+        name = 'false'
+      !ERROR: SELECT CASE statement value at this location overlaps with the below location
+      case (.true.)
+       name = 'true'
+      !ERROR: SELECT CASE statement value at this location overlaps with the below location
+      case (grade5)
+       name = 'true'
+   end select
+   print*, "Your grade is ", name
+
+   select case (grade2)
+      case (100:)
+         print*, "Excellent!"
+      case (:30)
+         print*, "Excellent!"
+      case (40)
+         print*, "Excellent!"
+      case (90)
+         print*, "Excellent!"
+      case (91:99)
+         print*, "Very good!"
+      !ERROR: SELECT CASE statement value at this location overlaps with the below location
+      case (81:90)
```

Comment:
I'll repeat my suggestion to tell the user at least one of the overlapping values. Also, I think "below" and "above" are confusing because the "below" location is referring to the location of the message, but the actual location of the source is above.

I suggest a pair of messages like this:
CASE value '90' matches a previous CASE statement
Previous CASE statement matching '90'

Reply:
I will update the error message to show the overlapping values.